### ISSUE TITLE
Memory Leak when using backward hooks with fmodel

### ISSUE THREAD

**Issue by vsieplus** [facebookresearch/higher#75]
Hi, I recently came across a memory error while attempting to use backward hooks with the `fmodel` obtained from `higher.innerloop_ctx`, with the default settings (copy_initial_weights, track_higher_grads True). Specifically, it appears that when backward hooks are registered to modules whose weight has `requires_grad=True` on `fmodel`, followed by a foward pass, loss computation, and a call to `diffopt.step(loss)`, these modules are not cleared upon exiting the context, or returning from a function containing the context. 

Here is an example of a dummy hook and a function with such a loop that causes the error:

```python

def placeholder_hook(module, grad_in, grad_out):
    pass

def run_higher_loop(model, optimizer, criterion):
    example = torch.tensor([0, 1, 2], dtype=torch.float).unsqueeze(0)
    target = torch.tensor([0], dtype=torch.float)

    with higher.innerloop_ctx(model, optimizer) as (fmodel, diffopt):
        # add placeholder backward hook to the copied embedding layer of fmodel
        for module in fmodel.modules():
            if isinstance(module, nn.Linear):
                module.weight.requires_grad = True
                module.register_backward_hook(placeholder_hook)

        output = fmodel(example)
        loss = criterion(output, target)
        diffopt.step(loss)

        # ...
```

When called with the below sample model, the fmodel's copied `nn.Linear` layer appears to persist:
```python
# simple binary classification model
class Net(nn.Module):
    def __init__(self, input_size, output_size):
        super().__init__()
        self.linear = nn.Linear(input_size, output_size, bias=False)

    def forward(self, examples):
        output = self.linear(examples)
        return output.squeeze(0)

model = Net(input_size=3, output_size=1)
optimizer = optim.SGD(model.parameters(), lr=0.01)    
criterion = nn.BCEWithLogitsLoss()

for i in range(4):
    mem_report(i)
    run_higher_loop(model, optimizer, criterion)
```

Here is the output of `mem_report`, which demonstrates that a copy of our linear layer (with input size 3, output size 1) stays in memory after each function call to `run_higher_loop`,
```
Memory report #0:
<class 'torch.nn.parameter.Parameter'> torch.Size([1, 3])


Memory report #1:
<class 'torch.Tensor'> torch.Size([1, 3])
<class 'torch.nn.parameter.Parameter'> torch.Size([1, 3])


Memory report #2:
<class 'torch.Tensor'> torch.Size([1, 3])
<class 'torch.Tensor'> torch.Size([1, 3])
<class 'torch.nn.parameter.Parameter'> torch.Size([1, 3])


Memory report #3:
<class 'torch.Tensor'> torch.Size([1, 3])
<class 'torch.Tensor'> torch.Size([1, 3])
<class 'torch.Tensor'> torch.Size([1, 3])
<class 'torch.nn.parameter.Parameter'> torch.Size([1, 3])
```

Some notes about the error

- The error occurs with nn.Embedding as well, and likely some others
- The error does not arise if the module.weight has `requires_grad=False`, and the backward hook is registered
- The error does not arise if the module.weight has `requires_grad=True`, and the backward hook is not registered
- The error does not arise in either case when a forward hook is registered
- The error never arises if we never call `diffopt.step(loss)`
- The definition of the hook itself does not affect the error (see above)
- Removing the hooks at the end of the function manually does not remove the error (i.e. storing a `handle` from `register_backward_hook` and later calling `handle.remove()`

A temporary fix is to manually delete the fmodel module weights which have the hooks before exiting the function. Adding the block below to the end of `run_higher_loop` removes the issue above:

```python
for module in fmodel.modules():
    if isinstance(module, nn.Linear):
        del module.weight
```

**Comment by egrefen** [comment]
Thanks for raising the issue and for the detailed diagnosis, @vsieplus! This is super helpful. I'll look into it as soon as I can (likely after NeurIPS rebuttals).

**Comment by brando90** [comment]
did you try doing `del fmodel`?

**Comment by brando90** [comment]
@egrefen sorry for the firect ping but has this been fixed?

**Comment by brando90** [comment]
related? https://github.com/facebookresearch/higher/issues/105 https://github.com/facebookresearch/higher/issues/97
